packages/xstate-wallet/src/store/memory-store.ts

```diff
@@ -115,51 +113,70 @@ export class MemoryStore implements Store {
   }
 
+  private async initializeChannel(
+    channelConstants: ChannelConstants
+  ): Promise<MemoryChannelStoreEntry> {
+    const addresses = channelConstants.participants.map(x => x.signingAddress);
+
+    const myIndex = addresses.findIndex(address => !!this._privateKeys[address]);
+    if (myIndex === -1) {
+      throw new Error("Couldn't find the signing key for any participant in wallet.");
+    }
+
+    const channelId = calculateChannelId(channelConstants);
+
+    // TODO: There could be concurrency problems which lead to entries potentially being overwritten.
+    this.setNonce(addresses, channelConstants.channelNonce);
+    const entry =
+      this._channels[channelId] || new MemoryChannelStoreEntry(channelConstants, myIndex);
+
+    this._channels[channelId] = entry;
+    return Promise.resolve(entry);
+  }
+
   public async createChannel(
     participants: Participant[],
     challengeDuration: BigNumber,
     stateVars: StateVariables,
     appDefinition = AddressZero
-  ): Promise<string> {
+  ): Promise<ChannelStoreEntry> {
     const addresses = participants.map(x => x.signingAddress);
 
     const myIndex = addresses.findIndex(address => !!this._privateKeys[address]);
     if (myIndex === -1) {
       throw new Error("Couldn't find the signing key for any participant in wallet.");
     }
 
-    const currentNonce = this.getNonce(addresses);
-    const channelNonce = currentNonce ? currentNonce.add(1) : bigNumberify(0);
-    this.setNonce(addresses, channelNonce);
+    const channelNonce = this.getNonce(addresses).add(1);
     const chainId = '1';
 
-    const channelId = getChannelId({
+    const entry = await this.initializeChannel({
       chainId,
-      channelNonce: channelNonce.toString(),
-      participants: addresses
+      challengeDuration,
+      channelNonce,
+      participants,
+      appDefinition
     });
-    this._channels[channelId] = new MemoryChannelStoreEntry(
-      {channelNonce, chainId, participants, appDefinition, challengeDuration},
-      myIndex
-    );
 
     // sign the state, store the channel
-    this.addState(channelId, stateVars);
+    this.signState(entry.channelId, stateVars);
 
-    return Promise.resolve(channelId);
+    return Promise.resolve(entry);
   }
 
-  private getNonce(addresses: string[]): BigNumber | undefined {
-    return this._nonces[this.nonceKeyFromAddresses(addresses)];
+  private getNonce(addresses: string[]): BigNumber {
+    return this._nonces[this.nonceKeyFromAddresses(addresses)] || bigNumberify(-1);
   }
 
   private setNonce(addresses: string[], value: BigNumber) {
+    if (value.lte(this.getNonce(addresses))) throw 'Invalid nonce';
+
     this._nonces[this.nonceKeyFromAddresses(addresses)] = value;
   }
 
   private nonceKeyFromAddresses = (addresses: string[]): string => addresses.join('::');
 
-  addState(channelId: string, stateVars: StateVariables) {
+  signState(channelId: string, stateVars: StateVariables) {
```

Comment:
Should this be `signAndAddState`?  From the name I would expect I need to call `signState` and then `addState` but it looks like `signState` adds it by calling `channelStorage.signAndAdd`?

Reply:
I don't think the store should ever sign a state and not add it: If I sign it, I should add it; otherwise, I risk giving someone else unilateral control over the channel. Likewise, if I sign a state, an honest wallet should always tell its peers; otherwise it is keeping control over the channel to itself, leading to broken protocols.

The API of `MemoryChannelStoreEntry` should be irrelevant, whether or not `signState` on the `Store` interface is called `signState` or `signAndAddState`. Do you prefer `signAndAddState` for clarity, or is `signState` good enough in light of this?

Reply:
I think I prefer `signAndAddState` (or some variant of that) for clarity.